### PR TITLE
Promote CSIPersistentVolume feature to GA

### PR DESCRIPTION
**Pull request by jsafrane**
**What this PR does / why we need it**:
We want CSIPersistentVolume GA in 1.13, https://github.com/kubernetes/features/issues/178

/milestone 1.13

CSI has many sub-features, this is only about the PersistentVolume part we consider stable enough. Inline volumes and block volumes and similar features have / will have their own alpha features and are / will be tracked separately.

**Special notes for your reviewer**:
Originally, I filled separate PRs for individual commits here, now I open this PR with everything on one place.

**Release note**:

```release-note
CSIPersistentVolume feature, i.e. PersistentVolumes with CSIPersistentVolumeSource, is GA. 
CSIPersistentVolume feature gate is now deprecated and will be removed according to deprecation policy.
```

/sig storage
/sig api-machinery
cc @saad-ali, @liggitt

### CODE REVIEW

**Comment by jsafrane** [comment]
cc @liggitt, I created one PR for all CSIPersistentVolume GA changes.

File: staging/src/k8s.io/api/storage/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/types.go]: since we know we intend to make this a union type in the future (either persistentVolumeName or inlineVolumeSource), we should resolve how an apiserver that doesn't yet know about inlineVolumeSource should validate this object.

Comment (jsafrane) [reply staging/src/k8s.io/api/storage/v1/types.go]: Current plan is to follow what worked in the past. For example, when `CSIPersistentVolume` feature was alpha and enabled, v1.PersistentVolumeSource.CSI was recognized. If someone turned off the feature or downgraded, the field became invisible and there could be PersistentVolumeSource without any item that does not pass validation. I don't remember anyone complaining

Is the process different now?

Comment (liggitt) [reply staging/src/k8s.io/api/storage/v1/types.go]: That was an alpha feature, and the "enable / populate data / disable" scenario was not supported.

This is introducing a v1 type with validation that will fail when objects with the planned inline volume source field are submitted. That seems like something we should get in place before promoting to GA.

Comment (jsafrane) [reply staging/src/k8s.io/api/storage/v1/types.go]: I don't understand. 

> This is introducing a v1 type with validation that will fail when objects with the planned inline volume source field are submitted

So was `v1.PersistentVolumeSource` - it's an union. We added CSI:

```
type PersistentVolumeSource struct {
...

    // CSI (Container Storage Interface) represents storage that handled by an external CSI driver (Beta feature).
    // +optional
    CSI *CSIPersistentVolumeSource
}
```

`PersistentVolumeSource.CSI` field was behind alpha feature. If there was a PV with CSI and the feature was disabled, such PV could not be validated. Nobody complained. In future "CSI inline volume PR", we will add a field to another union, `VolumeAttachmentSource`. Is there any difference that I miss?

**Review by saad-ali — APPROVED** [review]
Approved from SIG Storage side.

/lgtm
/approve

**Comment by msau42** [comment]
/hold

We should wait until we get the CSI 1.0 spec changes approved and merged before moving the feature to GA.

**Comment by msau42** [comment]
(Let's continue to review so that everything is ready to go)

**Comment by msau42** [comment]
As discussed in the k8s-CSI standup today, the CSI spec changes will be tight, but feasible for this release.  So let's continue to review and merge this PR and try to avoid any last minute surprises.

/hold cancel

**Comment by saad-ali** [comment]
@liggitt can we get API approval?
/assign @liggitt

**Comment by saad-ali** [comment]
@bertinatto can you rebase.

File: staging/src/k8s.io/api/storage/v1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/types.go]: it doesn't appear this object follows normal spec/status conventions:

* there is no status subresource
* the rest handler allows any part of the object to be modified via the primary resource (typically, there would be a primary resource strategy which would clear status in PrepareForCreate and preserve existing status in PrepareForUpdate, and a status strategy which would do the opposite, only allowing status to be updated on update
* generation is not incremented on update when the spec changes

I also don't see this PR registering this resource to be served under `storage.k8s.io/v1`:

https://github.com/kubernetes/kubernetes/blob/0aba5577412ea9b2fd89a0c0843e9b0c94e8848d/pkg/registry/storage/rest/storage_storage.go#L63-L83

Comment (msau42) [reply staging/src/k8s.io/api/storage/v1/types.go]: @liggitt iiuc, we need to do something like this https://github.com/kubernetes/kubernetes/blob/0aba5577412ea9b2fd89a0c0843e9b0c94e8848d/pkg/registry/core/rest/storage_core.go#L191 to add the subresource?

**Comment by msau42** [comment]
Do we need to update the in-tree callers to use v1 as part of this PR too?

**Comment by liggitt** [comment]
> Do we need to update the in-tree callers to use v1 as part of this PR too?

No, typically you wait a release to update callers

**Comment by msau42** [comment]
Can we really call this feature GA if nothing is actually calling the v1 api?

File: staging/src/k8s.io/api/storage/v1/types.go
Comment (msau42) [reply staging/src/k8s.io/api/storage/v1/types.go]: @bertinatto to help give you a head start, I went ahead and added the status subresource here: https://github.com/msau42/kubernetes/commit/e5ccf30d61b079d4a3d5a54777442139ad611187.  You can go ahead and pull it into your branch.

Unit tests are passing, but it still needs to be tested e2e.  I am running e2es now, which will test that current v1beta API did not break.

Comment (msau42) [reply staging/src/k8s.io/api/storage/v1/types.go]: Sorry, didn't get around to running the e2es, and I g2g.  The hostpath tests are run as part of pull, but GCE PD tests which really test VolumeAttachment are serial.  If you need help running them, ask @davidz627

**Comment by saad-ali** [comment]
>> Do we need to update the in-tree callers to use v1 as part of this PR too?
>>
> No, typically you wait a release to update callers

That's surprising. Why is that?

> Can we really call this feature GA if nothing is actually calling the v1 api?

@liggitt thoughts? This would be a blocker for CSI to GA in 1.13 if that were the case.

File: staging/src/k8s.io/api/storage/v1/types.go
Comment (gnufied) [reply staging/src/k8s.io/api/storage/v1/types.go]: I ran e2e btw and they passing. But - since we are still using beta version of `VolumeAttachment` objects, we aren't really testing new type via e2e yet.

**Comment by liggitt** [comment]
> > > Do we need to update the in-tree callers to use v1 as part of this PR too?
> > 
> > 
> > No, typically you wait a release to update callers
> 
> That's surprising. Why is that?
> 
> > Can we really call this feature GA if nothing is actually calling the v1 api?
> 
> @liggitt thoughts? This would be a blocker for CSI to GA in 1.13 if that were the case.

Switching all callers to v1 in the same release it is introduced makes HA rollouts more difficult (all new controllers must speak to a new API server, or they will make API calls to v1 APIs that don't exist yet and return 404s). If the rollout strictly upgrades API servers before the controllers that speak to them, that can work, but I've seen enough deployments that run HA apiserver+controllermanager machines, with the controllers speaking to arbitrary apiservers through a load balancer that I'd prefer to wait to switch the controllers to v1.

**Review by liggitt — COMMENTED** [review]
in preparation for switching to the v1 api in 1.14, make sure callers are not expecting to set status fields via create or update:

```diff
        attachment := &storage.VolumeAttachment{
                ObjectMeta: meta.ObjectMeta{
                        Name: attachID,
                },
                Spec: storage.VolumeAttachmentSpec{
                        NodeName: node,
                        Attacher: csiSource.Driver,
                        Source: storage.VolumeAttachmentSource{
                                PersistentVolumeName: &pvName,
                        },
                },
-               Status: storage.VolumeAttachmentStatus{Attached: false},
        }

        _, err = c.k8s.StorageV1beta1().VolumeAttachments().Create(attachment)
```

**Comment by liggitt** [comment]
> So it sounds like we can move the API in v1.13. And move the callers in v1.14. But the callers should be "silently migrated" in the next release, so they are an implementation detail. And so as far as the overall "Kubernetes CSI feature" is concerned, it can still be considered GA in 1.13?

Yes, availability of v1 API (especially one promoted from v1beta1 with no changes) means that the API is supported, and the things using the API have been tested to work properly and at scale.

File: pkg/registry/storage/volumeattachment/storage/storage_test.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/storage/storage_test.go]: Yeah that is accurate. Do you know if there is a way? Reading through testapi, it appears that currently there is no way we can run both tests in same run. It always runs only one version in same group, even if multiple versions are specified in the `KUBE_TEST_API_VERSIONS` .

Comment (gnufied) [reply pkg/registry/storage/volumeattachment/storage/storage_test.go]: I have manually ran the tests against both v1beta1 and v1 separately and it works. I am looking if CI is set to run unit tests against the versions separately if not, we will have to probably live with it?

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy.go]: fixed

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy.go]: fixed

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy.go]: fixed

File: pkg/registry/storage/volumeattachment/storage/storage_test.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/storage/storage_test.go]: okay fixed. I removed all versioning stuff from this file and added separate test in strategy_test.go which covers version behavior better.

File: pkg/registry/storage/volumeattachment/storage/storage_test.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/storage/storage_test.go]: fixed

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (msau42) [inline pkg/registry/storage/volumeattachment/strategy.go]: This is a lot of fields to remember to reset for probably every api type, and a lot of places to update if we add new fields to meta in the future.  Could we have some common "resetMeta" function that all api types could share?

File: pkg/registry/storage/volumeattachment/strategy_test.go
Comment (msau42) [inline pkg/registry/storage/volumeattachment/strategy_test.go]: The test name says BetaV1, but the test case is testing both v1 and v1beta1?

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy.go]: I would hope that a variant of commit that @liggitt linked will be merged in future in api-machinery repo but I wanted to avoid scope of this PR to not include apimachinery too. But I can indeed move this to a common util function.

File: pkg/registry/storage/volumeattachment/strategy_test.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy_test.go]: ha! I read that is "Test Beta API and v1 API". I know technically beta version is called betav1. I can perhaps rewrite this name to be `TestBetaAndV1StatusUpdate`  .

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (msau42) [reply pkg/registry/storage/volumeattachment/strategy.go]: Agree I think this could be done as a separate cleanup pr

File: pkg/registry/storage/volumeattachment/strategy_test.go
Comment (liggitt) [inline pkg/registry/storage/volumeattachment/strategy_test.go]: add a label, annotation, finalizer, and ownerRef as well and make sure they are ignored

**Comment by liggitt** [comment]
one comment on the status strategy unit test, then lgtm

**Comment by liggitt** [comment]
is there any additional validation we should do in Strategy#Validate for non-v1beta1 creation as part of promoting this to v1? things that come to mind:

* spec.Attacher is allowed to be arbitrary length and format... are there any restrictions on the value returned from `GetPluginName()` we should enforce here?
* spec.Source.PersistentVolumeName is not verified to be a valid PV name

once we ship v1, we won't reasonably be able to tighten that validation

**Comment by jsafrane** [comment]
Added validation for  spec.Attacher and spec.Source.PersistentVolumeName

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: keep in mind that we cannot change the v1beta1 API. additional validation like this we want to do needs to be done only on creation, and only when the API request is not via the v1beta1 API.

That info is available in the context passed to volumeAttachmentStrategy#Validate, the same way this PR is doing different behavior in PrepareForCreate for v1beta1

File: pkg/apis/storage/validation/validation.go
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: same comment about only adding this validation in volumeAttachmentStrategy#Validate, and only when the api request is not coming from the v1beta1 API

**Comment by liggitt** [comment]
needs a bazel update as well

File: pkg/registry/storage/volumeattachment/strategy_test.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy_test.go]: fixed

File: pkg/apis/core/validation/validation.go
Comment (jsafrane) [reply pkg/apis/core/validation/validation.go]: @liggitt, fixed. I squashed two validation commits into one and introduced `ValidateVolumeAttachmentV1` that's called only when v1 instance is created. I hope I got it right.

Comment (jsafrane) [reply pkg/apis/core/validation/validation.go]: If I understand it correctly, this is just a temporary measure, right? VolumeAttachments are re-created during each Kubernetes update (nodes get drained -> all volumes are detached -> all VolumeAttachments are deleted). At some point Kubernetes will create v1 objects only, so we can merge the validation functions.

Or shall we support 3rd party external attachers that are going to use v1beta1 API forever?

Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: once we have v1, we can deprecate v1beta1 and stop serving it after the deprecation period. until we have the ability to sweep and remove existing invalid objects, we will have to tolerate data that could have been persisted via v1beta1

File: pkg/apis/core/validation/validation.go
Comment (msau42) [inline pkg/apis/core/validation/validation.go]: does this have the same issue where you can't delete the Pod object after downgrading to a version where the feature is disabled?

Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: yes

Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: well, not the pod, since this is in PV, but the same issue, yes

Comment (gnufied) [reply pkg/apis/core/validation/validation.go]: thanks. I will pull your commit and update this branch.

File: pkg/apis/storage/validation/validation.go
Comment (gnufied) [reply pkg/apis/storage/validation/validation.go]: I think this was fixed by commit below.

**Comment by msau42** [comment]
This is ready for final review

File: pkg/registry/storage/volumeattachment/storage/storage.go
Comment (liggitt) [inline pkg/registry/storage/volumeattachment/storage/storage.go]: follow-up: please add a type assertion to ensure this implements enough to patch status:

`var _ = rest.Patcher(&StatusREST{})`

File: staging/src/k8s.io/api/storage/v1/generated.proto
Comment (liggitt) [inline staging/src/k8s.io/api/storage/v1/generated.proto]: follow-up: s/maybe/may be/ (in existing versions as well)

**Comment by liggitt** [comment]
please open a PR with the follow-ups so we don't forget, even if it sits until 1.14 opens:

* [extract utility method to reset objectmeta fields for status update](https://github.com/kubernetes/kubernetes/pull/69929#discussion_r233295428)
* [add a label, annotation, finalizer, and ownerRef to the status update unit test and make sure they are ignored](https://github.com/kubernetes/kubernetes/pull/69929#discussion_r233303732)
* [add a type assertion to ensure this implements enough to patch status](https://github.com/kubernetes/kubernetes/pull/69929#discussion_r233694415)
* [fix "maybe" typo in docs](https://github.com/kubernetes/kubernetes/pull/69929/files#r233696462)

File: pkg/registry/storage/volumeattachment/strategy.go
Comment (gnufied) [reply pkg/registry/storage/volumeattachment/strategy.go]: logged here - https://github.com/kubernetes/kubernetes/issues/71056

**Comment by gnufied** [comment]
@liggitt github issue for follow up items - https://github.com/kubernetes/kubernetes/issues/71056 . I have assigned it to both myself and @jsafrane . Whoever gets to it first.

**Pull request merged**
